src/xml_parser_main_metadata_together.c

```diff
@@ -336,132 +326,213 @@ int cr_xml_parse_main_metadata_together(const char *primary_path,
+    new_iter->primary_f = cr_open(primary_path, CR_CW_MODE_READ, CR_CW_AUTO_DETECT_COMPRESSION, &tmp_err);
     if (tmp_err) {
-        ret = tmp_err->code;
         g_propagate_prefixed_error(err, tmp_err, "Cannot open %s: ", primary_path);
         g_clear_error(&tmp_err);
-        goto out;
+        cr_PkgIterator_free(new_iter, err);
+        return NULL;
     }
-    filelists_f = cr_open(filelists_path, CR_CW_MODE_READ, CR_CW_AUTO_DETECT_COMPRESSION, &tmp_err);
+    new_iter->filelists_f = cr_open(filelists_path, CR_CW_MODE_READ, CR_CW_AUTO_DETECT_COMPRESSION, &tmp_err);
     if (tmp_err) {
-        ret = tmp_err->code;
         g_propagate_prefixed_error(err, tmp_err, "Cannot open %s: ", filelists_path);
         g_clear_error(&tmp_err);
-        goto out;
+        cr_PkgIterator_free(new_iter, err);
+        return NULL;
     }
-    other_f = cr_open(other_path, CR_CW_MODE_READ, CR_CW_AUTO_DETECT_COMPRESSION, &tmp_err);
+    new_iter->other_f = cr_open(other_path, CR_CW_MODE_READ, CR_CW_AUTO_DETECT_COMPRESSION, &tmp_err);
     if (tmp_err) {
-        ret = tmp_err->code;
         g_propagate_prefixed_error(err, tmp_err, "Cannot open %s: ", other_path);
         g_clear_error(&tmp_err);
-        goto out;
+        cr_PkgIterator_free(new_iter, err);
+        return NULL;
     }
 
     //TODO(amatej): In the future we could make filelists/other optional if there is a need for it. That would mean we
     //              should replace the last 0 in primary_parser_data_new depending on whether we have filelists or not.
-    primary_pd = primary_parser_data_new(newpkgcb_primary, &cbdata, pkgcb_primary, &cbdata, warningcb, warningcb_data, 0);
-    filelists_pd = filelists_parser_data_new(newpkgcb_filelists, &cbdata, pkgcb_filelists, &cbdata, warningcb, warningcb_data);
-    other_pd = other_parser_data_new(newpkgcb_other, &cbdata, pkgcb_other, &cbdata, warningcb, warningcb_data);
-
-    gboolean primary_is_done = 0;
-    gboolean filelists_is_done = 0;
-    gboolean other_is_done = 0;
-    cbdata.in_progress_count_primary = 0;
-    cbdata.in_progress_count_filelists = 0;
-    cbdata.in_progress_count_other = 0;
-    while (!primary_is_done || !filelists_is_done || !other_is_done) {
-        while ((cbdata.in_progress_count_primary <= cbdata.in_progress_count_filelists ||
-                cbdata.in_progress_count_primary <= cbdata.in_progress_count_other) &&
-               !primary_is_done)
+    new_iter->primary_pd = primary_parser_data_new(newpkgcb_primary, cbdata, pkgcb_primary, cbdata, warningcb, warningcb_data, 0);
+    new_iter->filelists_pd = filelists_parser_data_new(newpkgcb_filelists, cbdata, pkgcb_filelists, cbdata, warningcb, warningcb_data);
+    new_iter->other_pd = other_parser_data_new(newpkgcb_other, cbdata, pkgcb_other, cbdata, warningcb, warningcb_data);
+    return new_iter;
+}
+
+int cr_PkgIterator_parseNext(cr_PkgIterator *iter, cr_Package **package, GError **err) {
+    cr_CbData *cbdata = (cr_CbData*) &iter->cbdata;
```

Comment:
`iter->cbdata` already is a pointer.
```suggestion
    cr_CbData *cbdata = (cr_CbData*) iter->cbdata;
```